### PR TITLE
New Modeling Backend, BetaGeoModel, and PPD Metric

### PR DESCRIPTION
**Pull request by ColtAllen**
# Overview
The current[ autograd](https://github.com/HIPS/autograd) backend for `lifetimes` is no longer being actively developed, and fits models via maximum-likelihood estimation. MLE is an estimate of the posterior distribution mode (i.e., the peak of the bell curve) for each model parameter. It's fast, but considering only a single parameter value is akin to missing the forest for the trees and can be limiting in a number of ways.

In https://github.com/ColtAllen/lifetimes/issues/6, @juanitorduz provided a link to a notebook he wrote containing variants of `BetaGeoFitter` written in [pymc](https://github.com/pymc-devs/pymc), the premiere Python library for Bayesian statistical modeling. Bayesian methods can estimate the full posterior distribution of each parameter and confer many advantages for model tuning and interpretation, so I've adapted Juan's code into a new backend for `lifetimes` and a Beta-Geometric/NBD model class called `BetaGeoModel`. Best of all, [pymc](https://github.com/pymc-devs/pymc) has a very large and engaged developer community, ensuring the dependencies for this library will never be deprecated.

# New Feature Description

**UML diagram of the new modeling backend:**

models_uml

The new `BaseModel` object contains methods and attributes shared between all models and is also an abstract class, providing a template for future models as well as enforcing API standardization.

[In Bayesian modeling, the user specifies prior distributions for model parameters based on their own subjective intuition.](https://github.com/stan-dev/stan/wiki/Prior-Choice-Recommendations) Here's the stochastic dependency graph of what I came up with for `BetaGeoModel`:

betageo_graph

Parameter posteriors are estimated by the No-U-Turn Sampler (NUTS), a variation of the Hamiltonian Monte Carlo (HMC) sampling algorithm. The following link contains research papers links and interactive demos of these and similar algorithms in action:

[The Markov-chain Monte Carlo Interactive Gallery](https://chi-feng.github.io/mcmc-demo/)

For larger datasets (perhaps >30k rows) prior specifications become trivial because the data will overwhelm the subjectivity of the priors. However, the smaller the dataset, the more important prior selection becomes. These default model priors performed well in testing, but in a future PR I want to give users more options for tunability.

After a model has been fitted, it will contain an `idata` attribute of type `InferenceData` object, which can be plotted and analyzed with the [ArviZ](https://arviz-devs.github.io/arviz/examples/index.html) library. Here's an example slightly modified from a code excerpt in Juan's notebook:

```
import lifetimes as lt
import pymc as pm
import arviz as az

data_df = lt.load_cdnow_summary()

# Fit current lifetimes BetaGeo model
bg_mle = lt.BetaGeoFitter().fit(
    frequency=data_df['frequency'].values,
    recency=data_df['recency'].values ,
    T=data_df ['T'].values
    )

# Fit proposed new BetaGeo Bayesian model
bg_bayes = lt.BetaGeoModel().fit(data_df)

# Use ArviZ to plot posterior parameter distributions against the MLE estimates
axes = az.plot_trace(
    data=bg_bayes.idata,
    var_names=["a", "b", "alpha", "r"],
    lines=[(k, {}, [v]) for k, v in bg_mle.summary["coef"].items()],
    compact=True,
    backend_kwargs={
        "figsize": (12, 9),
        "layout": "constrained"
    },
)
fig = axes[0][0].get_figure()
fig.subtitle("BG/NBD Model Trace")
```
bg_nbd_arviz_plots

The existing user API for `lifetimes` remains fully intact; in fact the predictive methods for `BetaGeoModel` were copy/pasted right over from `BetaGeoFitter`! The coolest thing about a Bayesian approach is that repeated sample draws from the parameter posteriors will build an entire predictive probability distribution for any given customer, enabling prediction intervals and further analysis around customer behavior. Here's another slightly-modified example from Juan's notebook:

```
from matplotlib import pyplot as plt
import seaborn as sns

import lifetimes as lt

data_df = lt.load_cdnow_summary()

# Fit current lifetimes BetaGeo model and estimate p_alive
bg_mle = lt.BetaGeoFitter().fit(
    frequency=data_df['frequency'].values,
    recency=data_df['recency'].values ,
    T=data_df ['T'].values
    )

p_alive_mle = bg_mle.conditional_probability_alive(
    frequency=data_df['frequency'].values,
    recency=data_df['recency'].values ,
    T=data_df ['T'].values
    )

# Fit proposed new BetaGeo Bayesian model and estimate p_alive
bg_bayes = lt.BetaGeoModel().fit(data_df)
p_alive_bayes = bg_bayes.conditional_probability_alive()

# Plotting function to compare results
def plot_conditional_probability_alive(p_alive_bayes, p_alive_mle , idx, ax):
    sns.kdeplot(x=p_alive_sample[idx], color="C0", fill=True, ax=ax)
    ax.axvline(x=p_alive[idx], color="C1", linestyle="--")
    ax.set(title=f"idx={idx}")
    return ax

fig, axes = plt.subplots(
    nrows=3,
    ncols=3,
    figsize=(9, 9),
    layout="constrained"
)
for idx, ax in enumerate(axes.flatten()):
    plot_conditional_probability_alive(p_alive_sample, p_alive, idx, ax)

fig.subtitle("Conditional Probability Alive", fontsize=16)
```
bg_nbd_pymc_prob_alive_plots

### Collaborators: The plotting function defined in that code block would a great addition to the `plotting.py` module in a future PR.

**Please note these graphs cannot be recreated yet with the current version of `lt.BetaGeoModel().conditional_probability_alive()` because it only supports point estimates at this time. However, posterior sampling for predictions is a top priority for the next version release of this library.**

Instead of saving/loading the full model object via pickle files - which I consider a security risk because anything can be pickled, including malware - model persistence now comes in the form of a stripped-down JSON containing arrays for posterior parameter distributions and some metadata. During testing, the typical file size of this JSON was about 1.5 MB.

# Comparisons to `BetaGeoFitter`

A downside to this new backend is that full posterior estimation requires more time to fit a model than MLE. The tests I ran on the CDNOW dataset took about a minute to complete on 2.4k rows of data, whereas MLE converged in seconds. However, [pymc](https://github.com/pymc-devs/pymc) has experimental support for [jax](https://jax.readthedocs.io/en/latest/) just-in-time-compilation (JIT) and even GPUs, which could speed things up quite a bit and be a great future enhancement for `lifetimes`.

The mean values of the parameter posteriors, as well as the predictive outputs they generated, were comparable to that of `BetaGeoFitter` by 1-3 decimal places when fit to the same CDNOW dataset during testing. For well-behaved benchmarking datasets like CDNOW, MLE will outperform Bayesian estimates. However, real-world data is noisy and constantly influenced by unknown external factors, posing overfitting risks for point estimates of model parameters. Full posterior distributions provide a range of possible parameter values that vary every time the model runs, and repeated runs are loosely analogous to an ML modeling stack, affording greater resilience and explainability for the irreducible uncertainties of reality.

# Posterior Predictive Deviation Metric

Bayesian posterior predictive checks (PPCs) are usually a subjective visual exercise, comparing graphs of stochastic model outputs against observed data. However, an article on Medium inspired me to create a potential metric for PPCs based on the Wasserstein Distance, which I've dubbed the `posterior_predictive_deviation()`:

[Statistical Tests Won't Help You to Compare Distributions](https://towardsdatascience.com/statistical-tests-wont-help-you-to-compare-distributions-d829eefe418)

It can be interpreted as the number of standard deviations required to transform the distribution of model outputs into that of the observed data. The [Wasserstein Distance implementation in scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wasserstein_distance.html) only calculates distances between 1D arrays, but does allow optional weight vectors to be specified. During testing I weighed the Wasserstein Distance of customer frequency against recency because I wanted to consider both variables in this metric, but this is an uncharted new path for model interpretation and demands further analysis.

### After this PR is merged I intend to proceed with publishing an alpha release to PyPI. This new modeling backend is still incomplete and lacking adequate documentation, but it's a start. Functionality elsewhere in `lifetimes` remains unchanged.

### CODE REVIEW

**Comment by ColtAllen** [comment]
> @ColtAllen ! Amazing job! I will take a look at it next week (been busy these days) 🚀 !

Thanks, and that's great! I want to get this PR merged and the alpha release published to PyPI before Fri, 17-Jun, because I'm going on vacation afterward and won't be resuming work on `lifetimes` until July. However, if you find yourself in Mykonos 23-29 June let me know ✈

File: .gitignore [hunk 1]
```diff
@@ -75,3 +75,7 @@ benchmarks/images
 .amltmp
 *.amltmp 
 *.amltemp
+
+# lifetimes Testing Artifacts
+betageo_fitter.pkl
+exploded.csv
```
Comment (juanitorduz) [inline .gitignore:81]: Do we maybe want to add this into a `test/fixtures` directory so that we can run the test in a CI/CD ?

File: .gitignore [hunk 2]
```diff
@@ -36,7 +36,7 @@ pip-delete-this-directory.txt
 # Unit test / coverage reports
 htmlcov/
 .tox/
-.coverage
+.coverage.*
```
Comment (juanitorduz) [inline .gitignore:39]: There are still `.coverage` files in the main directory which should be removed.

File: lifetimes/models/__init__.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+        #self.model = self._model()
+
+        with self._model():
+            self.idata = pm.sample(
+                tune=tune,
+                draws=draws,
+                chains=4,
+                cores=len(psutil.Process().cpu_affinity()), 
+                target_accept=0.95,
```
Comment (juanitorduz) [inline lifetimes/models/__init__.py:90]: Can we add a `sample_kwargs` as an input for the function? Sometimes one would like to modify them, for example the `target_accept` value.

File: lifetimes/models/__init__.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+                return_inferencedata=True
+            )
+        
+        self.param_dict = self.idata.posterior.to_dict()
```
Comment (juanitorduz) [inline lifetimes/models/__init__.py:94]: I wonder if its better for the user to be able to use the `InferenceData` object itself as `xarray` is a very rich library.

File: lifetimes/models/beta_geo_model.py [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+        t_x = at.as_tensor_variable(recency)
+        T = at.as_tensor_variable(T)
+
+        x_zero = at.where(x>0, 1, 0)
+
+        # Refactored for numerical error
+        d1 = at.gammaln(r + x) - at.gammaln(r) + at.gammaln(a + b) + at.gammaln(b + x) - at.gammaln(b) - at.gammaln(a + b + x)
+        d2 = r * at.log(alpha) - (r + x) * at.log(alpha + t_x)
+        c3 = ((alpha + t_x)/(alpha + T))**(r+x)
+        c4 = a/(b+x-1)
```
Comment (juanitorduz) [inline lifetimes/models/beta_geo_model.py:133]: We should have a linter for code style right? Black? So that the spaces are consistent. `c4 = a / (b + x -1)`

Comment (juanitorduz) [reply lifetimes/models/beta_geo_model.py:133]: There are many places where we would need to fix a bit the style (e.g spaces). We can use black and flake8. These linters should be added in the CI/CD pipeline. I could support setting this up :)

File: lifetimes/utils.py [hunk 6]
```diff
@@ -687,3 +693,38 @@ def _concat2(*arrays):
     arrays = [np.array(ar) for ar in arrays]
     arrays = [ar.reshape((-1,1)) if ar.ndim < 2 else ar for ar in arrays]
     return np.concatenate(arrays, axis=-1)
+
+
+def posterior_predictive_deviation(obs_freq: npt.ArrayLike, gen_freq: npt.ArrayLike, obs_rec: npt.ArrayLike, gen_rec: npt.ArrayLike) -> float:
+    """ 
+    In lieu of a traditional posterior predictive check, calculate the standardized wasserstein distance of frequency, weighed by recency.
```
Comment (juanitorduz) [inline lifetimes/utils.py:700]: In view?

File: lifetimes/models/beta_geo_model.py [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+    def _model(self) -> pm.Model():
+
+        with pm.Model(name=f'{self.__class__.__name__}') as self.model:
+            # Priors for lambda parameters.
+            alpha_prior = pm.Weibull(name="alpha", alpha=1, beta=6)
+            r_prior = pm.Weibull(name="r", alpha=1, beta=1)
+
+            # Heirarchical pooling of hyperpriors for beta parameters.
+            phi_prior = pm.Uniform('phi', lower=0.0, upper=1.0)
+            kappa_prior = pm.Pareto('kappa',alpha=1,m=1.5)
```
Comment (juanitorduz) [inline lifetimes/models/beta_geo_model.py:62]: It would be nice to be able to allow the user to modify the priors, for example as in Bambi : https://bambinos.github.io/bambi/main/notebooks/radon_example.html

**Review by juanitorduz — CHANGES_REQUESTED** [review]
Hey! I had a quick look, as I was in a business trip this week, and I know you want feedback in order to iterate (and go to vacations 😎 !). Here Are somme comments:

## To Fix:

- The `pymc` package is not in the requirements so I can not run the tests. For example, when trying ` pytest -v tests/test_models.py ` I get:
```python
ImportError while importing test module '/Users/juanitorduz/Documents/lifetimes/tests/test_models.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../opt/anaconda3/envs/lifetimes_dev/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_models.py:12: in <module>
    import arviz as az
```

File: lifetimes/models/beta_geo_model.py (hunk 7 above)
Comment (juanitorduz) [reply lifetimes/models/beta_geo_model.py:62]: Ups! I just read this would be part of the second iteration 🙈 !

File: .gitignore (hunk 2 above)
Comment (ColtAllen) [reply .gitignore:39]: Yes; `.gitignore` was updated after these coverage files showed up. I'll have to delete them from the remote branch directly.

File: lifetimes/models/__init__.py (hunk 3 above)
Comment (ColtAllen) [reply lifetimes/models/__init__.py:90]: That's a great suggestion; I'll create an issue for it.

File: lifetimes/models/__init__.py (hunk 4 above)
Comment (ColtAllen) [reply lifetimes/models/__init__.py:94]: A few days after posting this PR I had a shower thought and came to the same conclusion. It's a simple modification that will enhance model interpretation considerably. I'll create an issue for this also.

File: .gitignore (hunk 1 above)
Comment (ColtAllen) [reply .gitignore:81]: This was copy/pasted straight over from the BG/NBD Time-Invariant Covariates PR in the first repo: https://github.com/CamDavidsonPilon/lifetimes/pull/342

In general I don't want to make any changes to the `fitters` module (or by extension, the tests for that module) because the new Bayesian `models` module will be replacing it, but I'll make an exception here because these external testing artifacts are disrupting test coverage reports.

I'll be merging this PR as-is because I have limited time this weekend before my vacation, but I'll create issues for this and anything else you encounter.

File: lifetimes/models/beta_geo_model.py (hunk 5 above)
Comment (ColtAllen) [reply lifetimes/models/beta_geo_model.py:133]: That would be great! I've been dragging my feet on setting up CI/CD due to evaluating the myriad of options available, but it's one of my goals.

File: lifetimes/models/beta_geo_model.py (hunk 7 above)
Comment (ColtAllen) [reply lifetimes/models/beta_geo_model.py:62]: Yes indeed! The extent of modification that Bambi supports would take some time to implement, but making the prior distribution parameters arguments for `BetaGeo._model` is straightforward.

File: lifetimes/utils.py (hunk 6 above)
Comment (ColtAllen) [reply lifetimes/utils.py:700]: Can you provide more detail? I'll need to merge this PR today, but I can still answer any questions you have and/or create an issue to implement any proposed changes afterward.

**Comment by ColtAllen** [comment]
Thanks @juanitorduz! I've replied to all of your review comments and will be creating work issues for the requested changes, because I need to get this PR merged today before I go on vacation.

The `requirements.txt` and other package files are actually leftovers from older versions of `lifetimes` which I'll be removing today before publishing the alpha release.

**Comment by juanitorduz** [comment]
Hey! Congrats on the alpha release! I'll try to give a more detail review and test it 🚀 ! Enjoy your time off!

**Comment by ColtAllen** [comment]
Thanks @juanitorduz! I'm back from vacation and have[ reorganized the work tasks into a tentative release roadmap.](https://github.com/ColtAllen/btyd/issues/16).

**Comment by zwelitunyiswa** [comment]
@ColtAllen Instead of a JSON, would an InferenceData object (i.e. Xarray) not make more sense to persist files? You would then get Arviz functions for free as well.

**Comment by ColtAllen** [comment]
> @ColtAllen Instead of a JSON, would an InferenceData object (i.e. Xarray) not make more sense to persist files? You would then get Arviz functions for free as well.

Hey @zwelitunyiswa, as of the Beta release [the full `InferenceData` object is now persisted as a JSON](https://github.com/ColtAllen/btyd/pull/41). If I can figure out [the issue with loading from Pandas DataFrames](https://github.com/ColtAllen/btyd/issues/40) then many more formats could be supported like CSV, Parquet & Feather, etc. In general I don't like using Pickle files because they can obscure malware and pose security risks.
